src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs

```diff
@@ -697,14 +697,35 @@ private ICollection<string> WarningsAsErrors
             }
         }
 
+        private ICollection<string> _warningsAsMessages;
+
+        /// <summary>
+        /// Contains all warnings that should be logged as errors.
+        /// Non-null empty set when all warnings should be treated as errors.
+        /// </summary>
+        private ICollection<string> WarningsAsMessages
+        {
+            get
+            {
+                // Test compatibility
+                if (_taskLoggingContext == null)
+                {
+                    return null;
+                }
+
+                return _warningsAsMessages ??= _taskLoggingContext.GetWarningsAsMessages();
+            }
+        }
+
         /// <summary>
         /// Determines if the given warning should be treated as an error.
         /// </summary>
         /// <param name="warningCode"></param>
-        /// <returns>True if WarningsAsErrors is an empty set or contains the given warning code.</returns>
+        /// <returns>True if the warning should not be treated as a message and WarningsAsErrors is an empty set or contains the given warning code.</returns>
         public bool ShouldTreatWarningAsError(string warningCode)
         {
-            if (WarningsAsErrors == null)
+            // Warnings as messages overrides warnings as errors.
+            if (WarningsAsErrors == null || (WarningsAsMessages != null && WarningsAsMessages.Contains(warningCode)))
```

Comment:
super-nit:

```suggestion
            if (WarningsAsErrors == null || WarningsAsMessages?.Contains(warningCode) == true)
```

Reply:
Also, does empty `WarningsAsMessages` mean that all warnings should be treated as messages or does only `WarningsAsErrors` have this behavior?

Reply:
Great question, I had the same when doing this PR. It doesn't look like it. `ShouldTreatWarningAsError` has this check:
```c#
            if (WarningsAsErrors != null)
            {
                // Global warnings as errors apply to all projects.  If the list is empty or contains the code, the warning should be treated as an error
                //
                if (WarningsAsErrors.Count == 0 || WarningsAsErrors.Contains(warningEvent.Code))
                {
                    return true;
                }
            }
```

whereas `ShouldTreatWarningAsMessage` checks like so:

```c#
            // This only applies if the user specified /nowarn at the command-line or added the warning code through the object model
            //
            if (WarningsAsMessages?.Contains(warningEvent.Code) == true)
            {
                return true;
            }
```

There's no check for the count being zero, so I don't think that rule applies.